src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs

```diff
@@ -18,29 +18,104 @@
+        private readonly IDisposable issueChangedServerEventSourcePublisher;
+        private readonly IDisposable taintServerEventSourcePublisher;
+        private readonly IThreadHandling threadHandling;
+        private bool disposed;
+        private CancellationTokenSource sessionTokenSource;
+
         [ImportingConstructor]
         public ServerSentEventSessionManager(
             IActiveSolutionBoundTracker activeSolutionBoundTracker,
             ISonarQubeService sonarQubeClient,
+            IServerSentEventPump eventPump,
             IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher,
-            ITaintServerEventSourcePublisher taintServerEventSourcePublisher)
+            ITaintServerEventSourcePublisher taintServerEventSourcePublisher,
+            IThreadHandling threadHandling)
+        {
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.sonarQubeClient = sonarQubeClient;
+            this.eventPump = eventPump;
+            this.issueChangedServerEventSourcePublisher = issueChangedServerEventSourcePublisher;
+            this.taintServerEventSourcePublisher = taintServerEventSourcePublisher;
+            this.threadHandling = threadHandling;
+            activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionChanged;
+        }
+
+        public void Dispose()
+        {
+            if (disposed)
+            {
+                return;
+            }
+
+            activeSolutionBoundTracker.SolutionBindingChanged -= OnSolutionChanged;
+            sessionTokenSource?.Cancel();
+            issueChangedServerEventSourcePublisher?.Dispose();
+            taintServerEventSourcePublisher?.Dispose();
+            disposed = true;
+        }
+
+        private void OnSolutionChanged(object sender, ActiveSolutionBindingEventArgs activeSolutionBindingEventArgs)
+        {
+            var bindingConfiguration = activeSolutionBindingEventArgs.Configuration;
+            var isInConnectedMode = !bindingConfiguration.Equals(BindingConfiguration.Standalone);
+
+            EndSession();
+
+            if (!isInConnectedMode)
+            {
+                return;
+            }
+
+            InitializeSessionAsync(bindingConfiguration.Project.ProjectKey).Forget();
+        }
+
+        private async Task InitializeSessionAsync(string projectKey)
+        {
+            await threadHandling.SwitchToBackgroundThread();
+
+            sessionTokenSource = new CancellationTokenSource();
+            var serverSentEventsSession =
+                await sonarQubeClient.CreateServerSentEventsSession(projectKey,
+                    sessionTokenSource.Token); //todo what kind of errors it throws? Sync with Rita
+
+            if (serverSentEventsSession == null)
+            {
+                return;
+            }
+
+            await eventPump.PumpAllAsync(serverSentEventsSession);
+        }
+
+        private void EndSession()
         {
```

Comment:
Is this enough?
What about the task from the previous session? `eventPump.PumpAllAsync` will be a thread that never finishes due to a while(true) loop. We need a way to terminate that thread (perhaps `while(!cancellationRequested)`)

Reply:
the session object should throw on cancellation and cause the loop to exit

Reply:
Hmm. We're making an assumption here but I don't have a better suggestion for now.

Reply:
W/o this assumption, we also couldn't assume that getNextEvent would not block forever